daemon/api_sideload_n_try.go

```diff
@@ -774,6 +799,42 @@ func installedSnapInfo(st *state.State, instanceName string) (*snap.Info, error)
 	return snapInfo, nil
 }
 
+// installedSnapInfoMatcher returns a function that looks for the components
+// associated snap in the set of installed snaps.
+func installedSnapInfoMatcher(st *state.State) func(string, naming.ComponentRef) (*snap.Info, *apiError) {
+	return func(instanceName string, cref naming.ComponentRef) (*snap.Info, *apiError) {
+		info, err := installedSnapInfo(st, instanceName)
+		if err != nil {
+			if errors.Is(err, state.ErrNoState) {
+				return nil, SnapNotInstalled(instanceName, fmt.Errorf("snap owning %q not installed", cref))
+			}
+			return nil, BadRequest("cannot retrieve information for %q: %v", instanceName, err)
+		}
+		return info, nil
+	}
+}
+
+func uploadedOrInstalledSnapInfoMatcher(st *state.State, snaps []sideloadSnapInfo) func(string, naming.ComponentRef) (*snap.Info, *apiError) {
+	return func(instanceName string, cref naming.ComponentRef) (*snap.Info, *apiError) {
+		for _, sn := range snaps {
+			// this is safe, since instance names are not supported when
+			// sideloading multiple blobs at once
+			if sn.info.RealName == instanceName {
+				return sn.info, nil
+			}
+		}
+
+		info, err := installedSnapInfo(st, instanceName)
+		if err != nil {
+			if errors.Is(err, state.ErrNoState) {
+				return nil, SnapNotInstalled(instanceName, fmt.Errorf("snap owning %q is neither installed nor provided to sideload", cref))
+			}
+			return nil, BadRequest("cannot retrieve information for %q: %v", instanceName, err)
+		}
+		return info, nil
```

Comment:
I considered this, but didn't want to make `installedSnapInfoMatcher` take a parameter that sets the error string (especially since it is a format string).

And I think it is good to include in the error string that we are considering snaps that uploaded as well as installed.